arangod/Replication2/ReplicatedLog/LogFollower.cpp

```diff
@@ -0,0 +1,413 @@
+                                              AppendEntriesErrorReason::WRONG_TERM);
+  }
+
+  // It is always allowed to replace the log entirely
+  if (req.prevLogEntry.index > LogIndex{0}) {
+    if (auto conflict =
+            algorithms::detectConflict(data._inMemoryLog, req.prevLogEntry);
+        conflict.has_value()) {
+      auto [reason, next] = *conflict;
+
+      LOG_CTX("5971a", DEBUG, _loggerContext)
+          << "reject append entries - prev log did not match: " << to_string(reason);
+      return AppendEntriesResult::withConflict(_currentTerm, req.messageId, next);
+    }
+  }
+
+  return std::nullopt;
+}
+
+auto replicated_log::LogFollower::appendEntries(AppendEntriesRequest req)
+    -> arangodb::futures::Future<AppendEntriesResult> {
+  MeasureTimeGuard measureTimeGuard{_logMetrics->replicatedLogFollowerAppendEntriesRtUs};
+
+  auto self = _guardedFollowerData.getLockedGuard();
+
+
+  {
+    // Preflight checks - does the leader, log and other stuff match?
+    // This code block should not modify the local state, only check values.
+    if (auto result = appendEntriesPreFlightChecks(self.get(), req); result.has_value()) {
+      return *result;
+    }
+
+    self->_lastRecvMessageId = req.messageId;
+  }
+
+  {
+    // Transactional Code Block
+    // This code removes parts of the log and makes sure that
+    // disk and in memory always agree. We first create the new state in memory
+    // as a copy, then modify the log on disk. This is an atomic operation. If
+    // it fails, we forget the new state. Otherwise we replace the old in memory
+    // state with the new value.
+    auto newInMemoryLog = self->_inMemoryLog.takeSnapshotUpToAndIncluding(req.prevLogEntry.index);
+
+    if (self->_inMemoryLog.getLastIndex() != req.prevLogEntry.index) {
+      auto res = self->_logCore->removeBack(req.prevLogEntry.index + 1);
+      if (!res.ok()) {
+        LOG_CTX("f17b8", ERR, _loggerContext)
+            << "failed to remove log entries after " << req.prevLogEntry.index;
+        return AppendEntriesResult::withPersistenceError(_currentTerm, req.messageId, res);
+      }
+    }
+
+    // commit the deletion in memory
+    static_assert(std::is_nothrow_move_assignable_v<decltype(newInMemoryLog)>);
+    self->_inMemoryLog = std::move(newInMemoryLog);
+  }
+
+  struct WaitForQueueResolve {
+    WaitForQueueResolve(Guarded<WaitForQueue>::mutex_guard_type guard, LogIndex commitIndex) noexcept
+        : _guard(std::move(guard)),
+          begin(_guard->begin()),
+          end(_guard->upper_bound(commitIndex)) {}
+
+    Guarded<WaitForQueue>::mutex_guard_type _guard;
+    WaitForQueue::iterator begin;
+    WaitForQueue::iterator end;
+  };
+
+  // Allocations
+  auto newInMemoryLog = self->_inMemoryLog.append(_loggerContext, req.entries);
+  auto iter = std::make_unique<InMemoryPersistedLogIterator>(req.entries);
+  auto toBeResolved = std::make_unique<std::optional<WaitForQueueResolve>>();
```

Comment:
If this code section is performance-critical, we can probably move these two lines with the allocations above, so that they are outside the mutex-guarded section.

Reply:
I don't think there's need for that. The leader only ever has a single `AppendEntriesRequest` in flight. So, modulo timeouts and packet loss, there is never any contention on this mutex.